### ISSUE TITLE
train_new_entity_type.py is inter-mixing doc and entity_offset data with BILOU format training data

### ISSUE THREAD

**Issue by analystanand** [explosion/spaCy#1003]
Bug in the [train new entity type](https://github.com/explosion/spacy/blob/master/examples/training/train_new_entity_type.py)

I am trying add new entity named **technology**  and i have annotated data in **BILOU** format to train with the help of [train new entity type](https://github.com/explosion/spacy/blob/master/examples/training/train_new_entity_type.py).

Originial example is trained with entity offset format but this can also be [trained in **BILOU** format](https://spacy.io/docs/usage/training-ner) as explained.

My Training Data in **BILOU** format  is same as given below
` train_data = [

        ('Play a key role in developing high - performance , scalable and fault - tolerant applications written in Java ',
        ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'U-technology']),

        ('Focus around building custom APIs for a variety of connected devises ',
         ['O', 'O', 'O', 'O', 'U-technology', 'O', 'O', 'O', 'O', 'O', 'O'])
    ]`

**Complete Training Script with BILOU format**
```
from __future__ import unicode_literals, print_function
import spacy
from spacy.gold import GoldParse
import random
from pathlib import Path


def train_ner(nlp, train_data, output_dir):
    # Add new words to vocab
    for raw_text, _ in train_data:
        doc = nlp.make_doc(raw_text)
        for word in doc:
            _ = nlp.vocab[word.orth]
        print(doc,len(doc))

    for itn in range(20):
        random.shuffle(train_data)
        for raw_text, entity_offsets in train_data:
            print(doc, "token length", len(doc), "length of entity", len(entity_offsets),id(doc))
            print(entity_offsets)
            # doc = nlp.make_doc(raw_text)  #position of doc object assignment is
            gold = GoldParse(doc, entities=entity_offsets) #loc is refefrenced before assignment
            doc = nlp.make_doc(raw_text)
            nlp.tagger(doc)
            loss = nlp.entity.update(doc, gold)
    nlp.end_training()
    if output_dir:
        nlp.save_to_directory(output_dir)



def main(model_name, output_directory=None):
    nlp = spacy.load(model_name)
    if output_directory is not None:
        output_directory = Path(output_directory)
    #
    train_data = [

        ('Play a key role in developing high - performance , scalable and fault - tolerant applications written in Java ',
        ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'U-technology']),

        ('Focus around building custom APIs for a variety of connected devises ',
         ['O', 'O', 'O', 'O', 'U-technology', 'O', 'O', 'O', 'O', 'O', 'O'])
    ]
    nlp.entity.add_label('technology')
    ner = train_ner(nlp, train_data, output_directory)

if __name__ == '__main__':

    output_directory = "example_model"
    main('en',output_directory)
```

The above example is throwing error which is given below
  ```
gold = GoldParse(doc, entities=entity_offsets) #loc is refefrenced before assignment
 File "spacy/gold.pyx", line 294, in spacy.gold.GoldParse.__init__ (spacy/gold.cpp:10834)
IndexError: list index out of range
```

Above issue could be fixed by changing  the position of 'doc = nlp.make_doc(raw_text)' to declare before gold parse object which is currently declared after gold parse object and this creates issue of **Local variable might be referenced before assignment**

--------------------------------------------------------------------------------------------------------------------------
**Complete Training Script with entity offset format**

The below example is working fine when training data is entity offset format.

```
from __future__ import unicode_literals, print_function
import spacy
from spacy.gold import GoldParse
import random
from pathlib import Path


def train_ner(nlp, train_data, output_dir):
    # Add new words to vocab
    for raw_text, _ in train_data:
        doc = nlp.make_doc(raw_text)
        for word in doc:
            _ = nlp.vocab[word.orth]
        print(doc,len(doc))

    for itn in range(20):
        random.shuffle(train_data)
        for raw_text, entity_offsets in train_data:
            print(doc, "token length", len(doc), "length of entity", len(entity_offsets),id(doc))
            print(entity_offsets)
            # doc = nlp.make_doc(raw_text)  #position of doc object assignment is
            gold = GoldParse(doc, entities=entity_offsets) #loc is refefrenced before assignment
            doc = nlp.make_doc(raw_text)
            nlp.tagger(doc)
            loss = nlp.entity.update(doc, gold)
    nlp.end_training()
    if output_dir:
        nlp.save_to_directory(output_dir)



def main(model_name, output_directory=None):
    nlp = spacy.load(model_name)
    if output_directory is not None:
        output_directory = Path(output_directory)

    train_data = [
         (
         'Play a key role in developing high - performance , scalable and fault - tolerant applications written in Java ',
         [(105,109,'technology')]),
    
         ('Focus around building custom APIs for a variety of connected devises ',
          [(29,33,'technology')])
     ]

    nlp.entity.add_label('technology')
    ner = train_ner(nlp, train_data, output_directory)


if __name__ == '__main__':

    output_directory = "example_model"
    main('en',output_directory)
```

## Your Environment

* Operating System: Ubuntu 14.04
* Python Version Used: 3.5.2
* spaCy Version Used: 1.8.0
* Environment Information: Pycharm

**Comment by irfan-zoefit** [comment]
There is an issue in BILUO entity tags.

```python
from __future__ import unicode_literals, print_function
from spacy.tokens.doc import Doc

import spacy
from spacy.pipeline import EntityRecognizer
from spacy.gold import GoldParse

train_data = [
        ('Focus around building custom Java for a variety of connected devises ',
         ['O', 'O', 'O', 'O', 'U-Technology', 'O', 'O', 'O', 'O', 'O', 'O']),
        ('I like Java',
         ['O', 'O', 'U-Technology']),
        ('They worked on Java',
         ['O', 'O', 'O', 'U-Technology']),
        ('Java is good.',
         ['U-Technology', 'O', 'O'])
    ]

nlp = spacy.load('en')
ner = EntityRecognizer(nlp.vocab, entity_types=['Technology'])

for raw_text, offsets in train_data:
    doc = Doc(nlp.vocab, list(raw_text))
    gold = GoldParse(doc, offsets)
    ner.update(doc, gold)

doc2 = nlp.make_doc('rats make good pets')
x = ner(doc2)

for word in doc2:
    print(word.text, ' ->> ', word.ent_iob_, word.ent_type_)
```

There should be a property for like **ent_biluo_** to identify the BILUO tags.

I'm getting wrong output although testing it on train data. it should work fine but it showing wrong output.
```shell
Java  ->>  B Technology
is  ->>  I Technology
good  ->>  I Technology
.  ->>  I Technology
```

**Comment by analystanand** [comment]
NER training script has been updated in latest release..
Try the given below script modified for **BILOU tag.**

For[ saving and loading](https://spacy.io/docs/usage/saving-loading) they  have provided standard system ,you can use that i am getting quite good results.


```
from __future__ import unicode_literals, print_function
import spacy
from spacy.gold import GoldParse
import random
from pathlib import Path


def train_ner(nlp, train_data, output_dir):
    # Add new words to vocab
    for raw_text, _ in train_data:
        doc = nlp.make_doc(raw_text)
        for word in doc:
            _ = nlp.vocab[word.orth]
        print(doc,len(doc))

    for itn in range(20):
        random.shuffle(train_data)
        for raw_text, entity_offsets in train_data:
            print(doc, "token length", len(doc), "length of entity", len(entity_offsets),id(doc))
            print(entity_offsets)
            # doc = nlp.make_doc(raw_text)  #position of doc object assignment is
            gold = GoldParse(doc, entities=entity_offsets) #loc is refefrenced before assignment
   
            nlp.tagger(doc)
            loss = nlp.entity.update(doc, gold)
    nlp.end_training()
    if output_dir:
        nlp.save_to_directory(output_dir)



def main(model_name, output_directory=None):
    nlp = spacy.load(model_name)
    if output_directory is not None:
        output_directory = Path(output_directory)
    
train_data = [
        ('Focus around building custom Java for a variety of connected devises ',
         ['O', 'O', 'O', 'O', 'U-Technology', 'O', 'O', 'O', 'O', 'O', 'O']),
        ('I like Java',
         ['O', 'O', 'U-Technology']),
        ('They worked on Java',
         ['O', 'O', 'O', 'U-Technology']),
        ('Java is good.',
         ['U-Technology', 'O', 'O'])
    ]
    nlp.entity.add_label('technology')
    ner = train_ner(nlp, train_data, output_directory)

if __name__ == '__main__':

    output_directory = "example_model"
    main('en',output_directory)
```

**Comment by irfan-zoefit** [comment]
Well I've updated the spacy from master branch and it still does not work. Which release you are talking about?

**Comment by honnibal** [comment]
@irfan-zoefit I think the mistake is that

```
gold = GoldParse(doc, offsets)
```

Should be

```
gold = GoldParse(doc, entities=offsets)
```

**Comment by irfan-zoefit** [comment]
@honnibal The issue is how you test that the tags are working properly? Above given examples only shows how to add the BILUO tags not how to verify that those tags are correctly mapped.

**Comment by irfan-zoefit** [comment]
I'm getting this error
```shell
  File "spacy/syntax/parser.pyx", line 318, in spacy.syntax.parser.Parser.update (spacy/syntax/parser.cpp:9653)
  File "spacy/syntax/ner.pyx", line 100, in spacy.syntax.ner.BiluoPushDown.preprocess_gold (spacy/syntax/ner.cpp:4960)
  File "spacy/syntax/ner.pyx", line 136, in spacy.syntax.ner.BiluoPushDown.lookup_transition (spacy/syntax/ner.cpp:5653)
KeyError: 'U-Technology'
```

and line is 

```python
    loss = nlp.entity.update(doc, gold)
```

**Comment by ines** [comment]
Quick update: we've since released spaCy v2.0, and updated [all training examples](https://github.com/explosion/spaCy/blob/master/examples/training/) for the new version. Usage should now be much simpler, and most examples include the expected output and additional documentation.

See [`train_new_entity_type.py`](https://github.com/explosion/spaCy/blob/master/examples/training/train_new_entity_type.py) and the step-by-step guide in the new documentation: https://spacy.io/usage/training#example-new-entity-type

For more details on training and updating the models, see the [new documentation](https://spacy.io/usage/training). There's now also a ["simple training style"](https://spacy.io/usage/training#training-simple-style) that lets you pass in text strings and annotation dictionaries (instead of `Doc` and `GoldParse` objects).
